server/rosetta/cosmos/client/client.go

```diff
@@ -0,0 +1,562 @@
+	"github.com/cosmos/cosmos-sdk/client/flags"
+	"github.com/cosmos/cosmos-sdk/client/tx"
+	"github.com/cosmos/cosmos-sdk/codec"
+	codectypes "github.com/cosmos/cosmos-sdk/codec/types"
+	"github.com/cosmos/cosmos-sdk/crypto/keys/secp256k1"
+	"github.com/cosmos/cosmos-sdk/server/rosetta"
+	"github.com/cosmos/cosmos-sdk/server/rosetta/cosmos/conversion"
+	"github.com/cosmos/cosmos-sdk/server/rosetta/services"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	grpctypes "github.com/cosmos/cosmos-sdk/types/grpc"
+	"github.com/cosmos/cosmos-sdk/types/tx/signing"
+	authclient "github.com/cosmos/cosmos-sdk/x/auth/client"
+	authsigning "github.com/cosmos/cosmos-sdk/x/auth/signing"
+	authtx "github.com/cosmos/cosmos-sdk/x/auth/tx"
+	auth "github.com/cosmos/cosmos-sdk/x/auth/types"
+	bank "github.com/cosmos/cosmos-sdk/x/bank/types"
+)
+
+// interface assertion
+var _ rosetta.NodeClient = (*Client)(nil)
+
+const tmWebsocketPath = "/websocket"
+
+// options defines optional settings for SingleClient
+type options struct {
+	interfaceRegistry codectypes.InterfaceRegistry
+	cdc               *codec.ProtoCodec
+}
+
+// newDefaultOptions builds the default options
+func newDefaultOptions() options {
+	cdc, ir := MakeCodec()
+	return options{
+		interfaceRegistry: ir,
+		cdc:               cdc,
+	}
+}
+
+// OptionFunc defines a function that edits single client options
+type OptionFunc func(o *options)
+
+// WithCodec allows to build a client with a predefined interface registry and codec
+func WithCodec(ir codectypes.InterfaceRegistry, cdc *codec.ProtoCodec) OptionFunc {
+	return func(o *options) {
+		o.interfaceRegistry = ir
+		o.cdc = cdc
+	}
+}
+
+// Client implements a single network client to interact with cosmos based chains
+type Client struct {
+	auth auth.QueryClient
+	bank bank.QueryClient
+
+	ir codectypes.InterfaceRegistry
+
+	clientCtx client.Context
+}
+
+func (c *Client) SupportedOperations() []string {
+	var supportedOperations []string
+	for _, ii := range c.ir.ListImplementations("cosmos.base.v1beta1.Msg") {
+		resolve, err := c.ir.Resolve(ii)
+		if err == nil {
+			if _, ok := resolve.(rosetta.Msg); ok {
+				supportedOperations = append(supportedOperations, strings.TrimLeft(ii, "/"))
+			}
+		}
+	}
+
+	supportedOperations = append(supportedOperations, rosetta.OperationFee)
+
+	return supportedOperations
+}
+
+func (c *Client) PreprocessOperationsToOptions(ctx context.Context, req *types.ConstructionPreprocessRequest) (options map[string]interface{}, err error) {
+	operations := req.Operations
+	if len(operations) > 3 {
+		return nil, rosetta.ErrInvalidRequest
+	}
+
+	msgs, err := conversion.ConvertOpsToMsgs(c.ir, operations)
+	if err != nil {
+		return nil, rosetta.WrapError(rosetta.ErrInvalidAddress, err.Error())
+	}
+
+	if len(msgs) < 1 || len(msgs[0].GetSigners()) < 1 {
+		return nil, rosetta.WrapError(rosetta.ErrInterpreting, "invalid msgs from operations")
+	}
+
+	memo, ok := req.Metadata["memo"]
+	if !ok {
+		memo = ""
+	}
```

Comment:
wouldn't it return an empty string if `!ok`? I'd just use `memo, _ := req.Metadata["memo"]`